src/System.IO.Pipelines/Pipe.cs

```diff
@@ -709,18 +789,41 @@ private void WriterCancellationRequested()
             TrySchedule(_writerScheduler, action);
         }
 
+        private void InvokeReaderCompletionCallbacks()
+        {
+            PipeCompletionCallbacks action;
+            lock (_sync)
+            {
+                action = _readerCompletion.GetCallbacks();
```

Comment:
GetCallbacks() should be called inline and only Execute() should be scheduled. Otherwise, the pipe could be reset before InvokeReaderCompletionCallbacks() gets run by the scheduler. If doing this requires allocating or running the callbacks inline, maybe we should just use a TCS.